File: apps/web/lib/track_link.py
"""
track_link.py — Python helper for building /r tracking redirect URLs.

Usage:
    from track_link import build_tracked_url

    url = build_tracked_url(
        "https://clinicaltrials.gov/study/NCT123",
        base_domain="https://trialchat.example.com",
        meta={
            "cta": "trial-card",
            "role": "caregiver",
            "intent": "trial_matching",
            "session_id": "abc123",
        },
    )
    # → "https://trialchat.example.com/r?to=https%3A%2F%2F...&meta=%7B...%7D"
"""

from __future__ import annotations

import json
import re
from typing import Literal, TypedDict
from urllib.parse import quote_plus, urlencode


class LinkMeta(TypedDict, total=False):
    cta: str
    role: Literal["user", "caregiver", "clinician"]
    intent: Literal["learn", "match", "trial_matching"]
    user_id: str
    session_id: str
    ref_page: str


_EXTERNAL_RE = re.compile(r"^(https?|tel|mailto):", re.IGNORECASE)


def build_tracked_url(
    raw_url: str,
    base_domain: str = "",
    meta: LinkMeta | dict | None = None,
) -> str:
    """
    Wrap an external URL in the /r tracking redirect.

    Args:
        raw_url:     The original destination URL (http/https/tel/mailto).
        base_domain: Optional domain prefix, e.g. "https://trialchat.example.com".
                     Leave empty to produce a root-relative path (/r?...).
        meta:        Optional metadata dict recorded alongside the click.

    Returns:
        Tracking URL string, or raw_url unchanged if it has no external protocol.
    """
    if not _EXTERNAL_RE.match(raw_url):
        return raw_url

    params: dict[str, str] = {"to": raw_url}
    if meta:
        params["meta"] = json.dumps(meta, separators=(",", ":"), ensure_ascii=False)

    query = urlencode(params, quote_via=quote_plus)
    return f"{base_domain}/r?{query}"

File: apps/web/lib/trackLink.ts
/**
 * trackLink.ts — frontend helper for building /r tracking redirect URLs.
 *
 * Usage:
 *   import { buildTrackedUrl } from "@/lib/trackLink";
 *
 *   <a href={buildTrackedUrl("https://clinicaltrials.gov/study/NCT123", {
 *     cta: "trial-enroll-btn",
 *     role: "caregiver",
 *     intent: "trial_matching",
 *     session_id: sessionId,
 *   })}>
 *     Enroll now
 *   </a>
 */

export interface LinkMeta {
  /** Label for the click-to-action source, e.g. "trial-card", "resource-link" */
  cta?: string;
  /** Current user role */
  role?: "user" | "caregiver" | "clinician";
  /** User intent captured during intake */
  intent?: "learn" | "match" | "trial_matching";
  /** Chat session user identifier, when available */
  user_id?: string;
  /** ChatKit / app session identifier */
  session_id?: string;
  /** Originating page pathname */
  ref_page?: string;
  /** Whether this click happened while the app was in test mode (?test=true) */
  is_test?: boolean;
  /** Allow arbitrary extension fields */
  [key: string]: unknown;
}

const EXTERNAL_RE = /^(https?:|tel:|mailto:)/i;

/**
 * Wraps an external URL in the /r tracking redirect.
 * Internal paths (no protocol) are returned unchanged.
 *
 * @example
 * buildTrackedUrl("https://clinicaltrials.gov/study/NCT123", { cta: "trial-card" })
 * // → "/r?to=https%3A%2F%2Fclinicaltrials.gov%2Fstudy%2FNCT123&meta=%7B%22cta%22%3A%22trial-card%22%7D"
 */
export function buildTrackedUrl(rawUrl: string, meta?: LinkMeta): string {
  if (!EXTERNAL_RE.test(rawUrl)) return rawUrl;

  const params = new URLSearchParams({ to: rawUrl });
  if (meta && Object.keys(meta).length > 0) {
    params.set("meta", JSON.stringify(meta));
  }
  return `/r?${params.toString()}`;
}
